design/fencing_mitigation.md

```diff
@@ -0,0 +1,34 @@
+# Problem Statement
+
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has healthy state) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as will as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy <= currentHealthy, we can disable fencing.
+
+# What is fencing?
+
+Once [MachineHealthCheck controller](https://github.com/openshift/machine-api-operator#machine-healthcheck-controller) detects that a node is `NotReady` (or some other configured condition), it will remove the associated `Machine` which will cause the node to be deleted. The `MachineSet` controller will then replace the `Machine` via the machine-api.
+
+# Why can't we use `PodDisruptionBudget`?
+
+This only applies to rook-ceph OSD pods.
+
+Any PDB with maxUnavailable >= 2 can cause storage downtime by having more than one node with OSDs down.
+
+Any PDB with maxUnavailable < 2 will completely block a node from being drained or fenced.
```

Comment:
Please explain. I think there are 2 unstated assumptions:
- There are 2 OSDs on the `NotReady` node.
- Storage for the OSDs is obtained from the local-volume provisioner (i.e., the PVs that the OSDs are using cannot be re-attached to other nodes)

Reply:
>     * There are 2 OSDs on the `NotReady` node.

 Yes that is an assumption. I'll include it. PDB would indeed protect from drains if there is only 1 OSD per node. I'll change the answer here to: "We cannot rely on fencing to drain the node first". I had upgrade on my mind where draining was also a factor.

>     * Storage for the OSDs is obtained from the local-volume provisioner (i.e., the PVs that the OSDs are using cannot be re-attached to other nodes)

In the baremetal case (where we use the local-volume provisioner), fencing will take to form of a reboot.